plugins/kvscheduler/api/kv_scheduler_api.go

```diff
@@ -262,7 +262,7 @@ type KVScheduler interface {
 type Txn interface {
 	// SetValue changes (non-derived) value.
 	// If <value> is nil, the value will get deleted.
-	SetValue(key string, value proto.Message) Txn
+	SetValue(key string, value proto.Message, groupings []string) Txn
```

Comment:
To provide more future-proof API the groups should be set via generic metadata map.

For example:
```go
SetValue(key string, val proto.Message, metadata map[string]string) Txn
// or
SetValue(key string, val proto.Message, metadata map[string]interface{}) Txn
```

Reply:
Both looks good to me.

To summarize and further generalize, I would propose that each graph node carries 4 fields, comprising what is currently known as "value":

```
IntendedConfig proto.message (currently hidden inside of the *LastUpdateFlag*)
ActualConfig proto.message
StateData proto.message (can be changed by a CRUD operation or through a notification)
MetaData map[string]interface{} / map[string][]string
```

At least one of the proto.Message must be non-nil for the node to be kept in the graph.

I would like to take out the indeded config from the LastUpdateFlag and make it an integral part of the "value". The flags should only carry run-time attributes of basic data-types, aiding the scheduling algorithm, but certainly not hiding proto messages.

`StateData` would consist of what is currently stored to `metadata` + some stuff that changes asynchronously, outside of CRUD operations, such as interface link status. The idea would be that state data can be edited by CRUD methods and via notifications. But since notification may want to alter only a subset of state data, we will probably have to delegate notification effects via callbacks, e.g.:
`PushSBNotification(key string, func (currentStateData proto.Message) newStateData proto.Message)`
Any alternative ideas are welcomed...

For most objects only configuration will be defined and `StateData` will be left as nil (e.g. routes, arps, dnats). Interfaces are the most complex items and will use both config and state data.
In few cases, such as microservices, the configuration is actually undefined and only  `StateData` will be received via `Retrieve` (renamed `Dump`) or through notifications.